pypesto/objective/finite_difference.py

```diff
@@ -0,0 +1,508 @@
+            delta = delta_val * unit_vec(dim=n_par, ix=ix)
+
+            if self.method == FD.CENTRAL:
+                fp = f_fval(x + delta / 2)
+                fm = f_fval(x - delta / 2)
+            elif self.method == FD.FORWARD:
+                fp = f_fval(x + delta)
+                fm = fval
+            elif self.method == FD.BACKWARD:
+                fp = fval
+                fm = f_fval(x - delta)
+            else:
+                raise ValueError("Method not recognized.")
+
+            grad[ix] = (fp - fm) / delta_val
+
+        return grad
+
+    def _hess_via_fd_fval(
+        self, x: np.ndarray, fval: float, **kwargs,
+    ) -> np.ndarray:
+        """Calculate 2nd order FD approximation to Hessian."""
+        # parameter dimension
+        n_par = len(x)
+
+        def f_fval(x):
+            """Short-hand to get a function value."""
+            return self.obj.call_unprocessed(
+                x=x, sensi_orders=(0,), mode=MODE_FUN, **kwargs)[FVAL]
+
+        hess = np.full(shape=(n_par, n_par), fill_value=np.nan)
+
+        # needed for diagonal entries at least
+        if fval is None:
+            fval = f_fval(x)
+
+        for ix1 in range(n_par):
+            delta1_val = self.get_delta_fun(par_ix=ix1)
+            delta1 = delta1_val * unit_vec(dim=n_par, ix=ix1)
+
+            # diagonal entry
+            if self.method == FD.CENTRAL:
+                f2p = f_fval(x + delta1)
+                fc = fval
+                f2m = f_fval(x - delta1)
+            elif self.method == FD.FORWARD:
+                f2p = f_fval(x + 2 * delta1)
+                fc = f_fval(x + delta1)
+                f2m = fval
+            elif self.method == FD.BACKWARD:
+                f2p = fval
+                fc = f_fval(x - delta1)
+                f2m = f_fval(x - 2 * delta1)
+            else:
+                raise ValueError("Method not recognized.")
+
+            hess[ix1, ix1] = (f2p + f2m - 2 * fc) / delta1_val ** 2
+
+            # off-diagonals
+            for ix2 in range(ix1):
+                delta2_val = self.get_delta_fun(par_ix=ix2)
+                delta2 = delta2_val * unit_vec(dim=n_par, ix=ix2)
+
+                if self.method == FD.CENTRAL:
+                    fpp = f_fval(x + delta1 / 2 + delta2 / 2)
+                    fpm = f_fval(x + delta1 / 2 - delta2 / 2)
+                    fmp = f_fval(x - delta1 / 2 + delta2 / 2)
+                    fmm = f_fval(x - delta1 / 2 - delta2 / 2)
+                elif self.method == FD.FORWARD:
+                    fpp = f_fval(x + delta1 + delta2)
+                    fpm = f_fval(x + delta1 + 0)
+                    fmp = f_fval(x + 0 + delta2)
+                    fmm = fval
+                elif self.method == FD.BACKWARD:
+                    fpp = fval
+                    fpm = f_fval(x + 0 - delta2)
+                    fmp = f_fval(x - delta1 + 0)
+                    fmm = f_fval(x - delta1 - delta2)
```

Comment:
I mean, there is the possibility to take the `f_fval(x+delta1 + delta2)` here and elsewhere (so drop the `/2` everywhere) and instead go for `(fpp - fpm - fmp + fmm) / (4 * delta1_val * delta2_val)`. This would be consistent with the diagonal elements...?

Reply:
Even though now I am confused, why there is no `4*` in the diagonals...?

Reply:
the step size should always be delta, i.e. from x - delta/2 to x + delta/2 (central), or from x to x + delta (forward). I think that should be consistent ...

Reply:
I think in the matlab implementation, it used a stepsize of delta for the off-diagonals, and delta/2 on the diagonals, which may be not the most consistent ... even though one would save a few evaluations when simultaneously computing gradients.